### PR TITLE
Wcf client dependency tracking

### PR DESCRIPTION
**Pull request by tomasr**
This PR implements code to emit `DependencyTelemetry` events for service calls done using the WCF client-side channel stack. Should work for HTTP, Net.TCP, and other bindings, and will include the name of the service operation being called.

Client-side telemetry tracking is disabled by default when installing the NuGet package and must be manually enabled.

### CODE REVIEW

File: WCF/Shared/Implementation/ClientTelemetryChannelBase.cs [hunk 1]
```diff
@@ -0,0 +1,361 @@
...
+        // telemetry implementation
+        private DependencyTelemetry StartOpenTelemetry(String method)
+        {
+            try
+            {
+                var telemetry = new DependencyTelemetry();
+                telemetry.Start();
+                telemetry.Type = DependencyConstants.WcfChannelOpen;
+                telemetry.Target = RemoteAddress.Uri.Host;
+                telemetry.Name = RemoteAddress.Uri.ToString();
```
Comment (SergeyKanzhelev) [inline WCF/Shared/Implementation/ClientTelemetryChannelBase.cs]: Name should have small cardinality. `Uri` likely will have arguments that should not be a part of name.

File: WCF/Shared/Implementation/ClientTelemetryChannelBase.cs [hunk 2]
```diff
@@ -0,0 +1,361 @@
...
+                telemetry.Data = ChannelManager.OperationMap.ContractType.Name;
```
Comment (SergeyKanzhelev) [inline WCF/Shared/Implementation/ClientTelemetryChannelBase.cs]: I think you need to exchange how you populate `Name` and `Data` here

File: WCF/Shared/Implementation/ClientTelemetryChannelBase.cs [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+                return telemetry;
+            } catch ( Exception ex )
```
Comment (SergeyKanzhelev) [inline WCF/Shared/Implementation/ClientTelemetryChannelBase.cs]: stylecop should have complained

File: WCF/Shared/Implementation/ClientTelemetryChannelBase.cs [hunk 4]
```diff
@@ -0,0 +1,361 @@
...
+        }
+
+        public IAsyncResult BeginOpen(TimeSpan timeout, AsyncCallback callback, object state)
+        {
+            WcfClientEventSource.Log.ChannelCalled(GetType().FullName, nameof(BeginOpen));
+            HookChannelEvents();
+            var telemetry = StartOpenTelemetry(nameof(BeginOpen));
+            try
+            {
+                return new OpenAsyncResult(InnerChannel, timeout, this.OpenCompleted, callback, state, telemetry);
```
Comment (SergeyKanzhelev) [inline WCF/Shared/Implementation/ClientTelemetryChannelBase.cs]: Is it possible to propagate dependency ID to the dependant WCF service for cross-component correlation?

File: WCF/Shared/Implementation/ClientTelemetryChannelBase.cs [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+            }
+
+            try
+            {
+                var telemetry = new DependencyTelemetry();
+                ChannelManager.TelemetryClient.Initialize(telemetry);
+                telemetry.Start();
+                telemetry.Type = DependencyConstants.WcfClientCall;
+                telemetry.Target = RemoteAddress.Uri.Host;
+                telemetry.Name = RemoteAddress.Uri.ToString();
```
Comment (SergeyKanzhelev) [inline WCF/Shared/Implementation/ClientTelemetryChannelBase.cs]: Same as before

File: WCF/Shared/Implementation/ClientTelemetryChannelBase.cs [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+            {
+                if ( error != null )
+                {
+                    telemetry.Success = false;
+                    telemetry.ResultCode = error.ToResultCode();
+                }
+                if ( response != null && response.IsFault )
+                {
+                    telemetry.Success = false;
+                    telemetry.ResultCode = "SoapFault";
```
Comment (SergeyKanzhelev) [inline WCF/Shared/Implementation/ClientTelemetryChannelBase.cs]: this is an excellent use of ResultCode! We have similar issue opened in SDK to track DNS error as a separate `ResultCode`. Thank you

File: WCF/Shared/Implementation/ProfilerWcfClientProcessing.cs [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+            if ( thisObj == null )
+            {
+                WcfClientEventSource.Log.NotExpectedCallback(0, nameof(OnEndInitializeEndpoint3), "thisObj == null");
+                return returnValue;
+            }
+            AddBehavior(((ChannelFactory)thisObj).Endpoint);
+            return returnValue;
+        }
+
+        private void AddBehavior(ServiceEndpoint endpoint)
```
Comment (SergeyKanzhelev) [inline WCF/Shared/Implementation/ProfilerWcfClientProcessing.cs]: Can you please rename to `AddIfNotExists`. This name confused me a little

**Comment by SergeyKanzhelev** [comment]
Thank you! And sorry for the delay.

File: WCF/Shared/Implementation/ClientTelemetryChannelBase.cs (hunk 4 above)
Comment (tomasr) [inline WCF/Shared/Implementation/ClientTelemetryChannelBase.cs]: Not when the channel is being opened. At that point, messages *could* be exchanged with the service, but they would be produced by the channels underneath us.

For example, if you're using WSHttpBinding with Message-based security and `negotiateServiceCredential` is enabled, the security channel will exchange infrastructure messages with the remote endpoint (doing 1 or more HTTP requests) but those will get handled by the WCF infrastructure and not make it to the service (the server side tracking won't see those requests, either).

We could potentially inject a second channel at the bottom of the stack (right before the encoding element) to do it, but not sure if it's worth it or not.

File: WCF/Shared/Implementation/ClientTelemetryChannelBase.cs (hunk 2 above)
Comment (tomasr) [inline WCF/Shared/Implementation/ClientTelemetryChannelBase.cs]: I can change those, no big deal. How does the portal display those, by default?

File: WCF/Shared/Implementation/ClientTelemetryChannelBase.cs (hunk 3 above)
Comment (tomasr) [inline WCF/Shared/Implementation/ClientTelemetryChannelBase.cs]: What's our StyleCop configuration? I think I might be missing that....

**Pull request merged**